src/sinks/util/service.rs

```diff
@@ -58,155 +59,84 @@ impl<L> ServiceBuilderExt<L> for ServiceBuilder<L> {
-    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
-    where
-        D: Deserializer<'de>,
-    {
-        struct UsizeOrAdaptive;
-
-        impl<'de> Visitor<'de> for UsizeOrAdaptive {
-            type Value = Concurrency;
-
-            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
-                formatter.write_str(r#"positive integer or "adaptive""#)
-            }
-
-            fn visit_str<E: de::Error>(self, value: &str) -> Result<Concurrency, E> {
-                if value == "adaptive" {
-                    Ok(Concurrency::Adaptive)
-                } else {
-                    Err(de::Error::unknown_variant(value, &["adaptive"]))
-                }
-            }
-
-            fn visit_i64<E: de::Error>(self, value: i64) -> Result<Concurrency, E> {
-                if value > 0 {
-                    Ok(Concurrency::Fixed(value as usize))
-                } else {
-                    Err(de::Error::invalid_value(
-                        Unexpected::Signed(value),
-                        &"positive integer",
-                    ))
-                }
-            }
-
-            fn visit_u64<E: de::Error>(self, value: u64) -> Result<Concurrency, E> {
-                if value > 0 {
-                    Ok(Concurrency::Fixed(value as usize))
-                } else {
-                    Err(de::Error::invalid_value(
-                        Unexpected::Unsigned(value),
-                        &"positive integer",
-                    ))
-                }
-            }
-        }
-
-        deserializer.deserialize_any(UsizeOrAdaptive)
-    }
-}
-
-pub trait ConcurrencyOption {
-    fn parse_concurrency(&self, default: &Self) -> Option<usize>;
-    fn is_none(&self) -> bool;
-    fn is_some(&self) -> bool {
-        !self.is_none()
-    }
-}
-
-impl ConcurrencyOption for Option<usize> {
-    fn parse_concurrency(&self, default: &Self) -> Option<usize> {
-        let limit = match self {
-            None => *default,
-            Some(x) => Some(*x),
-        };
-        limit.or(Some(5))
-    }
-
-    fn is_none(&self) -> bool {
-        matches!(self, None)
-    }
-}
-
-impl ConcurrencyOption for Concurrency {
-    fn parse_concurrency(&self, default: &Self) -> Option<usize> {
-        match self.if_none(*default) {
-            Concurrency::None => Some(5),
-            Concurrency::Adaptive => None,
-            Concurrency::Fixed(limit) => Some(limit),
-        }
-    }
-
-    fn is_none(&self) -> bool {
-        matches!(self, Concurrency::None)
-    }
-}
-
 /// Tower Request based configuration
-#[derive(Clone, Copy, Debug, Default, Deserialize, Serialize)]
+#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
 pub struct TowerRequestConfig<T: ConcurrencyOption = Concurrency> {
     #[serde(default)]
     #[serde(skip_serializing_if = "ConcurrencyOption::is_none")]
-    pub concurrency: T, // 5
+    pub concurrency: T, // 1024
```

Comment:
Just to confirm, is this right? I'm having a little trouble following how this is deserialized and defaulted. I see both `5` and `1024` in the context of `ConcurrencyOption` below.

Reply:
Where do you see the other 5? The path here is a little tortured but `concurrency.rs:30` is where the fixed concurrency value gets derived from, if I understand the logic properly.

Reply:
That would be https://github.com/timberio/vector/blob/0833d4dbeaa70d4d5fbd492b8ee9d4105e121c71/src/sinks/util/service/concurrency.rs#L19

Reply:
I indeed did not understand the logic here.

Reply:
Addressed in fbdc401

Reply:
👍 I'm still unclear why there are two places it is defaulted but they do seem ripe to be out-of-sync.